File: crates-backend/src/main/java/page/crates/service/AutoCategorizeService.java
package page.crates.service;

import page.crates.controller.api.autocategorize.AutoCategorizeExecuteRequest;
import page.crates.controller.api.autocategorize.AutoCategorizePreview;
import page.crates.controller.api.autocategorize.AutoCategorizeResult;

/**
 * Service for automatically categorizing user's music library into crates
 */
public interface AutoCategorizeService {

    /**
     * Preview what crates will be created (dry run)
     * Does not create any crates, just shows what would be created
     *
     * @return Preview of proposed crates
     */
    AutoCategorizePreview previewCategorization();

    /**
     * Execute auto-categorization
     * Creates 8-12 crates from user's library using multiple strategies
     *
     * @param request Optional request with pre-computed proposals from preview
     * @return Result with created crates
     */
    AutoCategorizeResult autoCategorize(AutoCategorizeExecuteRequest request);
}

File: crates-frontend/src/app/crate/shared/auto-categorize.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { AutoCategorizePreview, AutoCategorizeResult, CrateProposalDTO } from './model/auto-categorize.model';

@Injectable({
  providedIn: 'root'
})
export class AutoCategorizeService {

  constructor(private http: HttpClient) {}

  /**
   * Preview what crates will be created (dry run)
   */
  preview(): Observable<AutoCategorizePreview> {
    return this.http.get<AutoCategorizePreview>(
      `${environment.baseUri}/v1/auto-categorize/preview`
    );
  }

  /**
   * Execute auto-categorization (create crates)
   * @param proposals Optional pre-computed proposals from preview
   */
  execute(proposals?: CrateProposalDTO[]): Observable<AutoCategorizeResult> {
    const body = proposals ? { proposals } : {};
    return this.http.post<AutoCategorizeResult>(
      `${environment.baseUri}/v1/auto-categorize`,
      body
    );
  }
}
